### DISCUSSION TITLE
substring scripting command causes exception/reboot (*resolved*)

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#15724] [Q&A] [answered]
Not sure if it's just my syntax, but if I define two string variables, give a value to one of them then try to use the `sb()` command to pull off some characters, I get a crash.

```
>D
temp=0
hextest=0
chartest=0
str1="help"

>B
temp=0xBF3
hextest=hd(%temp%)
chartest=sb(str1 1 1)

>W
<center>hextest = %hextest%<br>
<center>temp = %temp%<br>
<center>chartest = %chartest%<br>
<center>str1 = %str1%<br>
```

Crash log shows: `"Exception":29,"Reason":"Exception","EPC":["4000e044","00000000","00000000"],"EXCVADDR":"00000000","DEPC":"00000000","CallChain":["40231aac","40293a15","40259ef0","4022f85f","40105563","401010d7","4010135f","402145b5","4024d271","401069c8","40251640","40202748","4024d271","402461da","40201306","40202630","40259ef0","40233475","40203e21","4022d3d3","401010d7","401010d7","4021e894","4010135f","401010d7","40256054","4021e894","4021e894","4022eaa0","4022eed0","4022ef4c"]}`

Also looking for some help concatenating hex bytes and nibbles to make a 12-bit decimal value. (Reading from subsequent 8-bit register cells)

*edit Looks like bitwise operators are my friend here. Need to bit shift then do some OR/AND masking. With a structure like the following:
<img width="909" alt="Screen Shot 2022-05-31 at 22 24 27" src="https://user-images.githubusercontent.com/79770016/171294071-3c304ed3-86f8-49e8-8e53-4945b2a5ea73.png">
I need to do something like `cell1=(CVR01<<4)|(CVR00&255)` and `cell2=(CVR02*16)|((CVR01>>4)&15)`

However, cell1 is (I guess) a 16-bit integer, so my shift isn't dropping the first four bits of CVR01 like I want. How do I make the first four bits 0? Some sort of modulus operation?

*edit2

Yes, the answer was %16 to just keep the 4 LSB, then shift left. So it's `cell1=((CVR01%16)<<8)|(CVR00&255)`

**Comment by sfromis** [comment, marked as answer]
Don't know about scripting, but it looks "suspect" that you init chartest with a numerical value, and then want to assign a substring to it, without any type of conversion.

**Reply by pkkrusty** [reply]
What's the normal syntax for initializing a string variable with a null value?

**Reply by sfromis** [reply]
Who told you that nulls are allowed? Empty string `""` may be the most feasible, to make the variable to be recognized as a string. Just my uninformed guess....

**Reply by gemu2015** [reply]
@sfromis is right. you must assign a substring to a  string.
so initialize  like this
 chartest=""
